src/esp/gfx/CubeMap.cpp

```diff
@@ -0,0 +1,399 @@
+      // TODO:
+      //{Mn::Shaders::Generic3D::ObjectIdOutput, objectIdAttachment}
+  });
+}
+void CubeMap::textureTypeSanityCheck(TextureType type,
+                                     const std::string& functionNameStr) {
+  switch (type) {
+    case TextureType::Color:
+      CORRADE_ASSERT(flags_ & Flag::ColorTexture,
+                     functionNameStr.c_str()
+                         << "instance was not created with color "
+                            "texture output enabled.", );
+      break;
+    case TextureType::Depth:
+      CORRADE_ASSERT(flags_ & Flag::DepthTexture,
+                     functionNameStr.c_str()
+                         << "instance was not created with depth "
+                            "texture output enabled.", );
+      break;
+    default:
+      CORRADE_INTERNAL_ASSERT_UNREACHABLE();
+      break;
+  }
+}
+
+Mn::GL::CubeMapTexture& CubeMap::getTexture(TextureType type) {
+  textureTypeSanityCheck(type, "CubeMap::getTexture():");
+  return *textures_[type];
+}
+
+std::string CubeMap::getTextureTypeFilenameString(TextureType type) {
+  switch (type) {
+    case TextureType::Color:
+      return std::string(".rgba");
+      break;
+    case TextureType::Depth:
+      return std::string(".depth");
+      break;
+    default:
+      CORRADE_INTERNAL_ASSERT_UNREACHABLE();
+      break;
+  }
+}
+Mn::PixelFormat CubeMap::getPixelFormat(TextureType type) {
+  switch (type) {
+    case TextureType::Color:
+      return Mn::PixelFormat::RGBA8Unorm;
+      break;
+    case TextureType::Depth:
+      return Mn::PixelFormat::R32F;
+      break;
+      /*
+      case TextureType::ObjectId:
+      return Mn::PixelFormat::R32UI;
+      */
+    default:
+      CORRADE_INTERNAL_ASSERT_UNREACHABLE();
+      break;
+  }
+}
+
+bool CubeMap::saveTexture(TextureType type,
+                          const std::string& imageFilePrefix) {
+#ifdef MAGNUM_TARGET_WEBGL
+  // because Mn::Image2D image = textures_[type]->image(...)
+  // requires desktop OpenGL
+  return false;
+#else
+  textureTypeSanityCheck(type, "CubeMap::saveTexture():");
+
+  Cr::PluginManager::Manager<Mn::Trade::AbstractImageConverter> manager;
+  Cr::Containers::Pointer<Mn::Trade::AbstractImageConverter> converter;
+  if (!(converter = manager.loadAndInstantiate("AnyImageConverter"))) {
+    return false;
+  }
+
+  std::string coordStrings[6] = {".+X", ".-X", ".+Y", ".-Y", ".+Z", ".-Z"};
+  for (int iFace = 0; iFace < 6; ++iFace) {
+    // TODO: use framebuffer.read() instead somehow?
+    // But the question is it has NOTHING to do with the framebuffer
+    // anymore at this moment.
+    Mn::Image2D image = textures_[type]->image(
+        convertFaceIndexToCubeMapCoordinate(iFace), 0, {getPixelFormat(type)});
+
+    std::string filename;
+    switch (type) {
+      case TextureType::Color: {
+        filename = imageFilePrefix + getTextureTypeFilenameString(type) +
+                   coordStrings[iFace] + std::string{".png"};
```

Comment:
LDR images are not really usable for IBL anyway, so rendering to a 8-bit framebuffer and saving as a PNG is not going to cut it in the long run. 

Which makes me wondering, wouldn't it just be easiest to have Magnum *properly* support OpenEXR? I was checking this just before Xmas due to #988 and the format is:

- 16-bit floats for color,
- 32-bit floats for Z, 
- understands mipmap,
- *and* has builtin support for cubemaps

so basically everything you need, and you would be able to store everything in a single file.

Reply:
>DR images are not really usable for IBL anyway

It still works, and the final result is just not that great.

So what you suggest here? 

Practical way is that I keep the way it is now, while you can develop the Magnum to support OpenEXR. After you have it, we can refactor the code.

Reply:
Yeah I'm just exploring the options. If OpenEXR isn't deemed useful, then great, less work for me; if it makes sense to have it / if it saves you or someone else extra work, then let's consider doing it.

Reply:
Actually, look at the pixel format I used for the color texture. RGBA8. Let us temporarily forget about the external HDR images, even internally, when I render the scene into such a texture, it is already has limitation for IBL because it is LBR. Correct? What is the correct pixel format then?

Reply:
> Actually, look at the pixel format I used for the color texture.

Yep, that's what I mentioned in #988 as well. For OpenEXR it'd be 16-bit floats, so RGBA16F.